nvflare/private/fed/client/client_runner.py

```diff
@@ -652,6 +655,32 @@ def end_run_events_sequence(self):
             self.fire_event(EventType.ABOUT_TO_END_RUN, fl_ctx)
             self.log_info(fl_ctx, "ABOUT_TO_END_RUN fired")
 
+            # check with all components for their readiness to end run
+            # use ConfigService to determine max wait time for readiness check:
+            #   the job config could define variable "end_run_readiness_timeout"
+            #   the user could define OS env var NVFLARE_END_RUN_READINESS_TIMEOUT
+            max_wait = ConfigService.get_float_var(
+                name="end_run_readiness_timeout", conf=self.config.config_data, default=5.0
+            )
+            check_start_time = time.time()
+            while True:
+                fl_ctx.remove_prop(FLContextKey.NOT_READY_TO_END_RUN)
+                self.log_info(fl_ctx, "Firing CHECK_END_RUN_READINESS ...")
+                self.fire_event(EventType.CHECK_END_RUN_READINESS, fl_ctx)
+                any_component_not_ready = fl_ctx.get_prop(FLContextKey.NOT_READY_TO_END_RUN, False)
+                if any_component_not_ready:
+                    if time.time() - check_start_time > max_wait:
+                        # we have waited too long
+                        self.log_warning(
+                            fl_ctx, f"quit waiting for component ready-to-end-run after {max_wait} seconds"
+                        )
+                        break
+                    else:
+                        time.sleep(0.5)
+                else:
+                    # all components are ready to end
+                    break
```

Comment:
This approach of checking if all the components are READY_TO_END_RUN could potentially cause  multiple fire_event(EventType.CHECK_END_RUN_READINESS, fl_ctx) calls. If there's any one component has a long processing task, the client_runner will fire this event to check every 0.5 seconds, which is a lot and maybe high CPU usage.

We can consider to change this to use the threading.Event() for waiting the component to finish tasking run. When firing the event to each component to handle, it creates a threading.Event(). Each component only sets the event if it finishes run. The firing event will wait for this event to be set with a timeout. With this approach, we don't have to call the fire_event(EventType.CHECK_END_RUN_READINESS, fl_ctx) multiple times.

Reply:
I'm not sure. First of all, the handler of this event should only check readiness but not do anything heavy. For example, the FedEventRunner (the real use case here) only checks to see whether it has any pending records. The processing of the pending records (the heavy processing) is not done in the event handler. 

Secondly, there can be any number of components/event handlers, managing threading.Event objects for them could be complex (and most of them are useless anyway since they are ready to end anyway).

Firing the events repeatedly shouldn't be a big deal here since checking for readiness should be a short process. Of course anyone could misuse this, but then this is true for all event types.

Reply:
Another thing is that the threading.Event based programming is non-trivial.  It may be harder to expect the application code (event handlers and their processing logic) to be able to handle it properly.